core/src/polling_monitor/mod.rs

```diff
@@ -0,0 +1,232 @@
+use std::collections::VecDeque;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use futures::stream;
+use futures::stream::StreamExt;
+use graph::cheap_clone::CheapClone;
+use graph::parking_lot::Mutex;
+use graph::prelude::tokio;
+use graph::slog::{debug, Logger};
+use tokio::sync::{mpsc, watch};
+use tower::{Service, ServiceExt};
+
+pub mod ipfs_service;
+
+/// Spawn a monitor that actively polls a service. Whenever the service has capacity, the monitor
+/// pulls object ids from the queue and polls the service. If the object is not present or in case
+/// of error, the object id is pushed to the back of the queue to be polled again.
+///
+/// The service returns the request ID along with errors or responses. The response is an
+/// `Option`, to represent the object not being found. `S::poll_ready` should never error.
+pub fn spawn_monitor<ID, S, E, Response: Send + 'static>(
```

Comment:
Would it make sense for the API to be `PollingMonitor::spawn` instead of a detached function `spawn_monitor`?

Perhaps you could even create a private `new` with these two lines:

```rust
    let queue = Arc::new(Mutex::new(VecDeque::new()));
    let (wake_up_queue, queue_woken) = watch::channel(());
```

Reply:
`PollingMonitor::spawn` is something I considered, but I wanted to try something different because `PollingMonitor` is just a handle to the task, and to me `Handle::spawn()` reads a bit backwards compared to `fn spawn() -> Handle`.